python/ray/serve/controller.py

```diff
@@ -345,6 +332,21 @@ async def run_control_loop(self) -> None:
             except Exception:
                 logger.exception("Exception updating application state.")
 
+            # Update the http proxy nodes set before updating the HTTP proxy states,
+            # so they are more consistent.
+            self._update_http_proxy_nodes()
+
+            # Don't update http_state until after the done recovering event is set,
+            # otherwise we may start a new HTTP proxy but not broadcast it any
+            # info about available deployments & their replicas.
+            if self.http_proxy_state_manager and self.done_recovering_event.is_set():
+                try:
+                    self.http_proxy_state_manager.update(
+                        http_proxy_nodes=self._http_proxy_nodes
+                    )
+                except Exception:
+                    logger.exception("Exception updating HTTP state.")
```

Comment:
why is this block moved?

Reply:
Dependency wise, http proxy state manager depends on the latest active nodes from deployment state manager so I think it's better to order the code following the dependency.

Reply:
with the new drain protocol, we shouldn't need to infer the draining nodes from deployment state manager anymore right? won't the ray scheduler/autoscaler decide for us?

Reply:
For draining nodes, you don't get it from deployment state manager but we still get active nodes from it. http proxy nodes is basically active_nodes - draining_nodes.